backoff.go

```diff
@@ -0,0 +1,181 @@
+package discovery
+
+import (
+	"math"
+	"math/rand"
+	"time"
+)
+
+type BackoffFactory func() BackoffStrategy
+
+type BackoffStrategy interface {
+	Delay() time.Duration
+	Reset()
+}
+
+// Jitter implementations taken roughly from https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+
+// Jitter must return a duration between min and max. Min must be lower than, or equal to, max.
+type Jitter func(duration time.Duration, min time.Duration, max time.Duration, rng *rand.Rand) time.Duration
+
+func FullJitter(duration time.Duration, min time.Duration, max time.Duration, rng *rand.Rand) time.Duration {
+	if duration <= min {
+		return min
+	}
+
+	normalizedDur := boundedDuration(duration, min, max) - min
+
+	return boundedDuration(time.Duration(rng.Int63n(int64(normalizedDur)))+min, min, max)
+}
+
+func NoJitter(duration time.Duration, min time.Duration, max time.Duration, rng *rand.Rand) time.Duration {
+	return boundedDuration(duration, min, max)
+}
+
+type randomizedBackoff struct {
+	min time.Duration
+	max time.Duration
+	rng *rand.Rand
+}
+
+func (b *randomizedBackoff) BoundedDelay(duration time.Duration) time.Duration {
+	return boundedDuration(duration, b.min, b.max)
+}
+
+func boundedDuration(d time.Duration, min time.Duration, max time.Duration) time.Duration {
+	if d < min {
+		return min
+	}
+	if d > max {
+		return max
+	}
+	return d
+}
+
+type attemptBackoff struct {
+	attempt int
+	jitter  Jitter
+	randomizedBackoff
+}
+
+func (b *attemptBackoff) Reset() {
+	b.attempt = 0
+}
+
+func NewFixedBackoffFactory(delay time.Duration) BackoffFactory {
```

Comment:
can we drop the `Factory` from these names? this is comically and very java-like.

Reply:
We can drop `Factory` from all the naming functions if you want, but it only really works because we currently have no non-Factory constructors. Would you want a name other than `BackoffFactory` for the type alias?

Btw, I noticed that Java-esque style while I was writing this and it didn't feel super great. However, we're returning functions that return backoff instances which is (like it or not) a factory. The reason we are using factories instead of just passing in the backoff is to allow for stateful backoffs (as is the case in decorrelated jitter and potentially other strategies).

Reply:
It's ok for the base type to have the factory name, but let's not be comical with the other names.
It is what it is, it doesn't have to be named like that too.
Also jave factories are slightly different, they don't return fucntions.

Reply:
done.

Reply:
btw if you have another more appropriate word for than factory for describing the thing that returns an a func for instantiating an object (creator, constructor, generator, etc.) feel free to suggest it.